components/post_editor.tsx

```diff
@@ -0,0 +1,418 @@
+    const html = (node.parentNode as Element).innerHTML;
+    return `${html.slice(0, innerHTMLStartOffset)}${replacement}${html.slice(innerHTMLEndOffset)}`;
+}
+
+function addNewTextNode(element: Element) {
+    const newNode = document.createTextNode('');
+    element.appendChild(newNode);
+    return newNode;
+}
+
+export default class PostEditor extends React.PureComponent<Props, {showingPlaceholder: boolean}> {
+    private selection: null | {start: number; end: number} = null;
+    private currentSelectionInvalid = false;
+
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {showingPlaceholder: (props.value || props.defaultValue || '') === ''};
+    }
+
+    private updateCurrentSelection = () => {
+        const selection = window.getSelection();
+        if (!selection || !this.refs.editor) {
+            return;
+        }
+
+        const range = selection.getRangeAt(0);
+        const clonedRange = range.cloneRange();
+        clonedRange.selectNodeContents(this.refs.editor as Element);
+        clonedRange.setEnd(range.endContainer, range.endOffset);
+        const end = clonedRange.toString().length;
+
+        clonedRange.setStart(range.startContainer, range.startOffset);
+        const start = end - clonedRange.toString().length;
+        this.selection = {start, end};
+    }
+
+    private setSelectionInEditor = () => {
+        if (!this.selection) {
+            return;
+        }
+
+        const range = document.createRange();
+        const sel = window.getSelection();
+        const childNodes = (this.refs.editor as Element).childNodes;
+        const {start, end} = getSelectionNodes(childNodes, this.selection.start, this.selection.end);
+        if (!sel || !start.node || !end.node) {
+            return;
+        }
+
+        range.setStart(start.node, start.offset);
+        range.setEnd(end.node, end.offset);
+        range.collapse(true);
+        sel.removeAllRanges();
+        sel.addRange(range);
+    }
+
+    private getEmoticons = (value: string) => {
+        const renderEmoji = renderEditorEmoji(this.props.emojiMap);
+        const tokens = new Map();
+        const withEmoticons = handleEmoticons(
+            value,
+            tokens,
+            {spaceRequiredForTextEmoticons: true, emojiRenderer: renderEmoji}
+        );
+
+        return {withEmoticons, tokens};
+    }
+
+    private updateEditorHtml = (value: string) => {
+        if (!this.refs.editor) {
+            return;
+        }
+
+        const {withEmoticons, tokens} = this.getEmoticons(value);
+        (this.refs.editor as Element).innerHTML = replaceTokens(withEmoticons, tokens);
+    }
+
+    private rearrangeNodes = (range: Range, currentNode: Node) => {
+        // If we've continued typing after creating an emoji then we need to remove the emoji span and
+        // add all the text into the last text node so we can check if there's still a valid emoji there
```

Comment:
I'm not sure if I understand this. Does this happen after a user has typed inside of an emoji's span? Maybe there's something I'm missing with what this function does?